쥬스메이커 [STEP 2] Groot, Borysarang

Groot-94 opened the pull request:
@protocorn93 
안녕하세요 콘 :) 😀
Step2 첫 PR 보냅니다.
이번에도 역시 많은 부분에서 고민해봤습니다. 리뷰 부탁드립니다! 

# 고민한점 🙋‍♂️
- **Navigation, Modality**
  - 재고를 수정하는 뷰로 넘어갈 때 Navigation방식으로 넘어갈지, Modal을 활용하여 present해줄 지 고민하였습니다. 화면이 단 두개뿐이라 흐름을 끊고 넘어간다는 점과, 음료를 주문하다가 재고를 변경해야해서 자연스럽게 넘어간다는 점 모두 해석에 따라 자연스럽다고 생각했습니다. 결론적으로 저희는 alert를 활용해서도 재고확인으로 넘어갈 수 있어야 해서 modal을 활용해서 화면전환을 구현했습니다.

- **closure 안에서 self를 써줘야 하는 이유**
  - NotificationCenter를 활용해서 UILabel을 업데이트해주는 함수를 호출하도록 구현했습니다. 다만 Observer를 등록하는 과정에서 클로저 내부에서 JuiceMakerViewController에 등록된 함수를 사용하려고 하니 self를 사용해야 했습니다. 왜 self를 사용해야 하는지에 대해 찾아보았는데 closure, 순환참조 등 가볍게 든 의문점인데 내용이 생각보다 많았습니다. 이 부분에 대해서 어떤 방향으로 공부를 해야하는지 고민을 했습니다.

- **Notifiction**
  - 과일의 재고를 나타내는 UILabel을 업데이트 해주는 과정에서 Notification을 사용했습니다. 이를 위해 과일 재고를 저장하는 FruitStore 인스턴스의 fruitStocks프로퍼티에 didset을 활용하여 NotificationCenter로 post해주도록 하였습니다. 하지만 didset에서 post해주는 방식은 최초에 NotificationCenter에 옵저버를 등록하기 전에(JuiceMakerViewController의 viewdidload 메서드에서 옵저버를 등록) post해주기 때문에 어플리케이션을 처음 실행하면 초기의 상태를 불러올 수 없었습니다. 그래서 저희는 초기에 FruitStock을 가져오는 함수를 구현하여 해결했지만, Notification을 활용하여 이를 해결 할 수 있는지 확인중에 있습니다. 

- **Button의 Action을 하나로 묶고 어떤 Button이 눌렸는지 확인하는 방법**
  - 이번에 기본적으로 제공된 UI에는 총 7개의 쥬스를 생성하는 버튼이 있습니다. 저희는 이 7개의 버튼이 모두 JuiceMaker의 takeOrder함수를 호출하기 때문에 모든 버튼이 하나의 IBAction을 공유하여 사용하고 어떤 버튼이 눌렸는지만 구분해주면 코드가 깔끔해질 것이라고 생각했씁니다. 다만 이 버튼을 구분하는 과정에서 어떤게 맞는지 고민을 했습니다. 
처음에는 `RestorationID`라는 속성이 버튼에 있길래 ID라고 표현한 것이 Identifier처럼 사용해도 되는것이라 생각하여 이를 활용해볼까 했지만, 이 속성은 다른 목적을 위해(어플리케이션이 임의의 종료를 당했을 때 이를 복구하기 위해 사용되는 것이라고 이해했습니다) 사용되는것 같았습니다. 두번째로 AccessibilityIdentifier속성을 활용하는 것이었는데 이 속성은 목적에는 맞았지만 String타입이어서 switch 조건절에서 String으로 너무 많이 비교하는것 같아서 보류하였습니다.
최종적으로 그냥 IBAction의 매개변수인 sender자체를 비교하였는데, 이렇게 여러 버튼들을 하나의 IBAction에서 처리할 때 어떻게 다루는지 궁금했습니다. 

- **StoryBoard에 등록된 여러 UILabel을 가져오는 방법**
  - 재고가 변경되면 FruitStore인스턴스의 fruitStock 프로퍼티를 가져와서 label을 업데이트 해주도록 구현하였습니다. 재고가 변경되면 storyboard에 등록된 stock들을 가져와서 각 과일의 재고를 업데이트 해주도록 했는데 5개나 되는 Label들을 일일이 `bananaStockLabel.text = stock`과 같이 업데이트 해주는게 맞는가 고민하였습니다. 
 
# UML  
[ClassDiagram]
![](https://i.imgur.com/6SLfIZP.png)
[Flowchart]
![](https://i.imgur.com/cJtopsy.jpg)

Groot-94 committed:
style: Update button position

yusw10 committed:
style: Update button property

Groot-94 committed:
add: Create FruitStoreViewController

- update naming to JuiceMakerViewController from ViewController

yusw10 committed:
feat: Add FruitStore BackButton

- add navigation bar button to backward
- add segue to navigation barbutton

Groot-94 committed:
feat: Create @IBOutlet, @IBAction, implement function

- create FruitStockLabel
- create FruitjuiceButton
- implement orderFruitJuice function

yusw10 committed:
refactor: Refactoring orderFruitJuice IBAction

- refactoring return type of JuiceMaker sturct
- refactoring orderFruitJuice function to use alertResult function
- add juiceMaker typeProperty in JuiceMakerViewController
- add alertResult function to debug orderFruitJuice

Groot-94 committed:
feat: Implement alertResult function

yusw10 committed:
feat: Add presenting Modal of FruitStoreViewController

- add presentModalView Function
- delete okAction, noAction's closure
- fix preferredStyle of alert

Groot-94 committed:
feat: Create Notification

- create Notification post fruitInventory didset of FruitStore
- create Notification addObserver viewDidLoad function of JuiceMakerViewController

yusw10 committed:
feat: Add fruit Stock function

- add NotificationCenter body for addObserving
- add updateFruitInventroy function for updating stockLabel text

Groot-94 committed:
feat: Add requestCurrentStock function

- refactor: update fruitInventory naming
- refactor: update  updateFruitInventory naming
- refactor: update  inventory naming

yusw10 committed:
feat: Fix alert information message

- add rawValue in FruitJuice
- delete print function
- modify alertResult function to fix alert message

Groot-94 committed:
refactor: Update naming

yusw10 committed:
refactor: Refactoring function and enum

- add stockLabel's identifier
- add Fruit enumeration's rawValue
- fix takeOrder function's style
- refactoring updateFruitsStockLabels function

Groot-94 committed:
fix: Update make function  bug

- Fixed a bug where, when making fruit juice using two fruits, the other fruit would decrease if only one fruit was insufficient

Groot-94 committed:
docs: Update README.md

yusw10 committed:
docs: Update README.md

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/FruitStoreViewController.swift:
```diff
@@ -0,0 +1,19 @@
+        super.viewDidLoad()
+    }
+
+    @IBAction func pressBackBarButton(_ sender: UIBarButtonItem) {
```
`@IBAction`이나 추후 사용하게 되실 `@objc` 메서드들도 `private`으로 작성되면 좋을 것 같아요.

Groot-94 replied:
수정 했습니다! 
[3d755b1](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/3d755b142dfbb606e899df2bd19aed7beb4bdd35)

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/FruitStoreViewController.swift:
```diff
@@ -0,0 +1,19 @@
+
+import UIKit
+
+class FruitStoreViewController: UIViewController {
```
저는 보통 상속되지 않을 클래스는 `final` 키워드를 붙여요. 지금 당장은 아니더라도 추후에 `final` 키워드에 대해 학습해보시면 좋을 것 같아요. 왜 붙이는지, 붙였을 때 어떤 이점이 있는지! 이를 학습하다보면 swift에 대해서도 보다 깊게 학습해볼 수 있을거에요!

yusw10 replied:
말씀 주신대로 `final` 키워드를 상속되지 않을 클래스(JuiceMakerViewController, FruitStoreViewController, FruitStore)에 지정해주었습니다.
final에 대해 조금 찾아보았는데 제가 생각한 이점은 다음과 같다고 생각했습니다.
1. final을 붙이면 스태틱 디스패치로 작동하기 때문에 내부 메서드들이 런타임시에 조금 더 성능적으로 이득이다.
2. 코드를 보는 다른 개발자로 하여금 이 클래스를 상속받는 하위 클래스는 없다는것이 명확해진다.
이 외에 사용하는 목적이 추가적으로 있을까요?
[cc7d219](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/cc7d219ac15d9a90b77eb95cb77587c4768dbb14)

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,101 @@
+    @IBOutlet weak var bananaJuiceButton: UIButton!
+    @IBOutlet weak var pineappleJuiceButton: UIButton!
+    @IBOutlet weak var kiwiJuiceButton: UIButton!
+    @IBOutlet weak var mangoJuiceButton: UIButton!
```
마찬가지로 접근제어자에 대해서 고민해주시길 바래요~! 접근제어자에 대한 코멘트는 여기까지 할게요!

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,101 @@
+                return
+            }
+            self.updateFruitsStockLabels(currentStock)
+        }
```
옵저버를 등록하는 코드도 별도의 메서드로 분리해보시는건 어떨까요? 메서드가 한 줄 짜리 코드만 포함할지라도 보다 명확한 메서드 네이밍으로 의도를 더 명확하게 전달할 수 있어요! 그리고 메서드의 추상화 레벨이 맞다면 가독성도 더 좋아져요! 

이런 추상화 레벨과 같은 개념에 있어서는 "클린 코드" 라는 책도 함께 읽어보시는걸 추천해요!

protocorn93 replied:
`Notification.Name("fruitsStock")`은 별도의 namespace로 분리해서 관리하는게 이후 유지보수에도 도움이 되요! 

```swift
enum FruitStoreNotificationName { 
  static let fruitsStockDidChanged = Notification.Name("furitsStockDidChanged")
}
```

Groot-94 replied:
수정 했습니다! 


- enum을 사용한 NotificationName [5211543](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/5211543b80e9dd4878704982feba30da306661d9)

- 옵저버 함수분리 [5559b01](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/5559b01351711874d92fd94d983ae9100c1e2538)

protocorn93 commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -5,10 +5,10 @@
+    case banana = "banana"
+    case kiwi = "kiwi"
+    case pineapple = "pineapple"
+    case mango = "mango"
```
rawValue가 String이라면 기본적으로 case 명이 그대로 rawValue로 사용되요! 그렇기 때문에 rawValue를 모두 지정해줄 필요 없어요! rawValue가 달라져야하는 case만 지정해주면 되요!

Groot-94 replied:
이 부분은 정말 몰랐습니다!
수정했습니다!
[5211543](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/5211543b80e9dd4878704982feba30da306661d9)

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,101 @@
+        }
+        for (fruit, value) in unwrappedStock {
+            modifyFruitStockLabel(fruit.rawValue, value)
+        }
```
이후에 고차함수를 학습하시게 되면 이를 활용해 수정해보실 수 있을 것 같아요.

yusw10 replied:
코멘트 주신대로 해당 부분 고차함수인 `compactMap`활용해서 과일 재고를 수정하는 메서드를 불러오도록 했씁니다.
다만 이부분을 고차함수로 작성하다보니 `compactMap`의 반환값을 사용하지 않아 wildCard로 처리해주었는데 이렇게 작성하는것이 말씀주신 고차함수 사용의 의도와 맞는지 잘 모르겠습니다.
[d1cefcd](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/d1cefcd8489bc68bfcbb98f4545d32745ed24440)

protocorn93 replied:
네 이부분은 추가로 코멘트 남겼어요!

protocorn93 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,19 +8,22 @@ import Foundation
 // 쥬스 메이커 타입
 struct JuiceMaker {
-    private let fruitStore = FruitStore(initialAmount: 10)
+    private let fruitStore: FruitStore = FruitStore(initialAmount: 10)
```
타입을 추가로 명시한 이유가 궁금하네요!

Groot-94 replied:
이 부분은 이전 기존에 전역변수들에 대해서 타입을 명시해줬기 때문에 동일하게 타입을 명시했습니다.
리뷰를 보고 collection같은 타입에 대해선 타입을 명시 해줘야 하지만, 클래스나 구조체에 대한 타입과 같이 구분이 가능하면 해주지 않아도 되겠다는 생각을 하게되었습니다.
지워주는 방향으로 수정했습니다.
[f4f5ff9](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/f4f5ff90d99945535c92e7d1996277e528f4853a)

protocorn93 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,19 +8,22 @@ import Foundation
+    private let fruitStore: FruitStore = FruitStore(initialAmount: 10)
     
-    func takeOrder(_ fruitJuice: FruitJuice) {
+    func takeOrder(_ fruitJuice: FruitJuice) -> (FruitJuice, Bool) {
```
tuple보다는 보다 명확한 네이밍을 갖는 객체를 활용하는게 좋을 것 같아요.

protocorn93 replied:
Swift의 Result 타입을 학습해보시고 이를 활용하면 조금 더 가독성 좋고 깔끔한 코드를 작성할 수 있을 것 같아요

yusw10 replied:
말씀주신대로 ResultType 활용해서 작성해보았습니다!
찾아보니 비동기 함수의 completion handler에서 에러를 처리할 때(네트워크 모듈을 사용하는 예제를 참고했습니다!) 사용하는게 주 목적인 것 같은데, 그 외에 do - catch문 대신 가독성을 위해서도 사용해도 괜찮은걸까요?
[4a72e54](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/4a72e547fb864a9c0629c7e8518538c3dfce958f)

protocorn93 replied:
네네~ 그럼요! completionHandler에서만 사용하라는 법은 없죠! 타입 네이밍과 타입 자체가 Error을 내포하기 때문에 보다 명확한 의미를 전달할 수 있어요!

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,101 @@
+            alert.addAction(noAction)
+            present(alert, animated: true)
+        }
+    }
```
메서드 안에서 result 값에 따라 alert이 분기되는데 각각을 별도의 메서드로 빼보시는건 어떨까요? 가독성이 더 좋아질 것 같아요

Groot-94 replied:
분리하도록 수정했습니다!

[d85c2cd](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/d85c2cdd3c915654a76f79bec728833fda961847)

하지만, 최종적으로 수정하는 과정에서 alertResult 함수의 필요성을 느끼지 못해서 alertResult 함수는 제거하는 방향으로  수정했습니다.

[f4f5ff9](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/f4f5ff90d99945535c92e7d1996277e528f4853a)

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,101 @@
+        guard let stock = stock else {
+            return
+        }
+        for uiLabel in allStockLabels where uiLabel.accessibilityIdentifier == fruit {
+            uiLabel.text = String(stock)
+        }
```
이 코드 역시 추후 고차함수를 학습하게 되시면 수정해보시는걸 권장드려요

yusw10 replied:
위의 메서드와 마찬가지로 filter 사용하여 수정해보았습니다.
[d1cefcd](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/d1cefcd8489bc68bfcbb98f4545d32745ed24440)

protocorn93 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,19 +8,22 @@ import Foundation
-            guard let completedFruitJuice = try fruitStore.make(fruitJuice) else {
-                print("재고가 부족합니다.")
-                return 
+            guard try fruitStore.make(fruitJuice) != nil else {
+                return (fruitJuice, false)
             }
-            print("\(completedFruitJuice)가 완성되었습니다.")
         } catch JuiceMakerError.invalidOrder {
             print("잘못된 음료 요청.")
         } catch {
             print(error)
         }
+        return (fruitJuice, true)
```
에러 핸들링은 최종적으로 사용하는 ViewController에서 처리해주는건 어떨까요? 에러에 따라 alert 문구도 달라질 수 있고 에러 핸들링 역시 달라질 수 있을 것 같은데 말이죠!

yusw10 replied:
코멘트 대로 에러 처리를 뷰컨트롤러단에서 처리하도록 throw를 한번 더 하게 만들었고, 위의 코멘트대로 ResultType을 사용하면서 가독성을 높이는데 주의하면서 기능분리를 시도해 보았습니다.
[4a72e54](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/4a72e547fb864a9c0629c7e8518538c3dfce958f)

protocorn93 commented on JuiceMaker/JuiceMaker/View/Base.lproj/Main.storyboard:
```diff
@@ -347,16 +517,22 @@
                     </view>
-                    <navigationItem key="navigationItem" id="g4W-fY-l7r"/>
+                    <navigationItem key="navigationItem" id="g4W-fY-l7r">
+                        <barButtonItem key="leftBarButtonItem" title="뒤로가기" id="NoE-Jg-fht">
```
모달로 올라오는 화면이기에 "뒤로가기"보단 "닫기"가 더 좋은 네이밍일 것 같아요.

Groot-94 replied:
수정했습니다!
[f4f5ff9](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/f4f5ff90d99945535c92e7d1996277e528f4853a)

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,101 @@
+    
+    func modifyFruitStockLabel(_ fruit: String, _ stock: Int?) {
+        let allStockLabels: [UILabel] = [strawberryStockLabel, bananaStockLabel, pineappleStockLabel, kiwiStockLabel, mangoStockLabel]
+        guard let stock = stock else {
+            return
+        }
```
guard 문을 제일 위에 위치시키는게 어떨까요? guard를 통한 unwrapping은 가급적이면 메서드의 최상단에 위치하는게 코드를 읽을 때 흐름을 덜 해칠수 있어요!

Groot-94 replied:
해당 guard 문은 고차함수를 사용하면서 제거했습니다 ㅎ
[d1cefcd](https://github.com/yagom-academy/ios-juice-maker/pull/232/commits/d1cefcd8489bc68bfcbb98f4545d32745ed24440)

protocorn93 reviewed (changes requested):
두 분 모두 고생하셨어요! 👏👏 쉽지 않은 스탭인데 잘 구현해주셨어요! 

## 고민한 점 

### Navigation, Modality

이 부분은 확실히 앱의 기능과 사용성에 따라 달라질 수 있을 것 같아요. 기능의 흐름에도 많은 영향을 미칠 수 있는 부분인데요. 이것에 대해선 HIG 문서를 한번 읽어보시면 조금은 감을 잡으실 수 있을거라 생각해요. 

- [Modality](https://developer.apple.com/design/human-interface-guidelines/ios/app-architecture/modality/) 문서
- [Navigation](https://developer.apple.com/design/human-interface-guidelines/ios/app-architecture/navigation/) 문서 


### closure 안에서 self를 써줘야 하는 이유

네 클로저! 이 개념 역시 상당히 어려운 개념이고 정말 iOS 개발자라면 늘 마주치는 녀석이죠! 방향성의 시작은 기본 문서라고 생각해요. 먼저 [Swift 공식 문서](https://docs.swift.org/swift-book/LanguageGuide/Closures.html)부터 천천히 읽어보시는걸 추천해드려요. 

### Button의 Action을 하나로 묶고 어떤 Button이 눌렸는지 확인하는 방법

하나의 IBAction 메서드로 소화하려면 현재 sender로 버튼을 비교하는 방법도 하나의 방법이에요! 그게 아니라면 버튼별 IBAction 메서드를 만들어주는 방법도 있겠죠! 


### StoryBoard에 등록된 여러 UILabel을 가져오는 방법

이 문제는 현재 구현 방법과 관련이 있는데요. 재고가 변경될 때마다 모든 과일재고를 넘겨주기보단 변경이 필요한 재료에 대해서 남은 재료만을 userInfo를 통해 전달해주면 변경이 필요한 label을 명확하게 알 수 있기에 이에 대한 고민은 해결 할 수 있을 것으로 보여져요! 

예를 들어 딸기 10, 바나나 10개가 있는 상태에서 딸기, 바나나 각각 2개가 사용됐다면 딸기:8, 바나나:8 이렇게 건내줄거고 그렇다면 명확히 딸기레이블, 바나나레이블에 접근해서 값을 변경해주는거죠!

Groot-94 committed:
refactor: Hiding functions, properties

yusw10 committed:
refactor: Refactoring ViewController and FruitStore Class

- add final keyword

Groot-94 committed:
add: NotificationName enum file

- replace magic number with symbolic constant
- remove rawValue of Fruit enum

Groot-94 committed:
Merge branch 'step2' of https://github.com/Groot-94/ios-juice-maker into step2

yusw10 committed:
refactor: Abstracting add observer

- seperate addObserver function of viewdidload

Groot-94 committed:
refactor: Abstracting alertResult

- seperate alertResult function

yusw10 committed:
- refactoring orderFruitJuice to use Result Type

- refactoring takeOrder function to use Result type
- renaming error naming of JuiceMakerError
- refactoring bindingStock function's throw
- refactoring orderFruitJuice function
- seperate AlertResult to alertResult and alertInvalidAcceess and alertfailureOfFruitJuice

Groot-94 committed:
refactor: Abstracting post

- seperate post function
- modified to pass only the information of the changed fruitstock

yusw10 committed:
- Refactoring Function to use high order function

- refactoring modifyFruitStockLabel function to use high order function
- refactoring updateFruitsStockLabels function to use high order function

Groot-94 committed:
refactor: Refactoring naming, type  declaration, storyboard

yusw10 commented:
@protocorn93 
콘! 두번쨰 STEP2 PR 코멘트 달아주신것들 수정 완료해서 다시 요청 드립니다!
잘부탁드립니당👏

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -115,20 +114,14 @@ final class JuiceMakerViewController: UIViewController {
-        }
-        for (fruit, value) in unwrappedStock {
-            modifyFruitStockLabel(fruit.rawValue, value)
+        _ = stock?.compactMap { (key: Fruit, value: Int) in
+            modifyFruitStockLabel(key.rawValue, value)
```
콜렉션 타입을 순회하면서 새로운 콜렉션을 만들어내는 것이 아니기에 map/compactMap은 사용할 필요는 없어보여요. 
(이 둘의 차이도 학습해보시길 권장드려요) 

단순히 순회하는 것이라면 forEach로도 충분해보이네요
